File: scripts/generate_dashboard.py
#!/usr/bin/env python3
"""
generate_dashboard.py

A script to generate SUBMODULE_DASHBOARD.md, outlining the status, current commit,
and branch of all submodules in the Omni-Workspace.
"""

import os
import subprocess
import datetime

def run_cmd(cmd, cwd=None):
    try:
        result = subprocess.run(cmd, cwd=cwd, check=True, text=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        return result.stdout.strip()
    except subprocess.CalledProcessError:
        return "Unknown"

def get_submodules():
    output = run_cmd(["git", "config", "--file", ".gitmodules", "--get-regexp", "path"])
    if output == "Unknown" or not output:
        return []

    paths = []
    for line in output.splitlines():
        parts = line.split(" ", 1)
        if len(parts) == 2:
            paths.append(parts[1])
    return paths

def get_repo_status(path):
    if not os.path.exists(os.path.join(path, ".git")):
        return "Not Initialized", "N/A", "N/A"

    branch = run_cmd(["git", "rev-parse", "--abbrev-ref", "HEAD"], cwd=path)
    commit = run_cmd(["git", "rev-parse", "--short", "HEAD"], cwd=path)

    # Check if working tree is clean
    status_output = run_cmd(["git", "status", "--porcelain"], cwd=path)
    status = "Clean" if not status_output else "Dirty"

    return branch, commit, status

def main():
    dashboard_file = "SUBMODULE_DASHBOARD.md"
    submodules = get_submodules()

    with open(dashboard_file, "w") as f:
        f.write("# Omni-Workspace Submodule Dashboard\n\n")
        f.write(f"*Last updated: {datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')}*\n\n")

        if not submodules:
            f.write("No submodules detected in `.gitmodules`.\n")
            return

        f.write("| Submodule | Branch | Commit | Status |\n")
        f.write("| --- | --- | --- | --- |\n")

        for sm in sorted(submodules):
            branch, commit, status = get_repo_status(sm)
            f.write(f"| `{sm}` | `{branch}` | `{commit}` | {status} |\n")

    print(f"Generated {dashboard_file}")

if __name__ == "__main__":
    main()

File: backend-go/generate_dashboard.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"sort"
	"strings"
	"time"
)

type DashboardEntry struct {
	Submodule string `json:"submodule"`
	Branch    string `json:"branch"`
	Commit    string `json:"commit"`
	Status    string `json:"status"`
}

type DashboardResponse struct {
	Message string `json:"message"`
	Success bool   `json:"success"`
}

func getDashboardSubmodules(repoPath string) ([]string, error) {
	cmd := exec.Command("git", "config", "--file", ".gitmodules", "--get-regexp", "path")
	cmd.Dir = repoPath
	out, err := cmd.Output()
	if err != nil {
		return nil, nil
	}

	var paths []string
	lines := strings.Split(strings.TrimSpace(string(out)), "\n")
	for _, line := range lines {
		parts := strings.SplitN(line, " ", 2)
		if len(parts) == 2 {
			paths = append(paths, parts[1])
		}
	}
	return paths, nil
}

func getRepoStatus(repoPath, smPath string) (string, string, string) {
	fullPath := repoPath + "/" + smPath

	if _, err := os.Stat(fullPath + "/.git"); os.IsNotExist(err) {
		return "N/A", "N/A", "Not Initialized"
	}

	cmdBranch := exec.Command("git", "rev-parse", "--abbrev-ref", "HEAD")
	cmdBranch.Dir = fullPath
	outBranch, err := cmdBranch.Output()
	branch := "Unknown"
	if err == nil {
		branch = strings.TrimSpace(string(outBranch))
	}

	cmdCommit := exec.Command("git", "rev-parse", "--short", "HEAD")
	cmdCommit.Dir = fullPath
	outCommit, err := cmdCommit.Output()
	commit := "Unknown"
	if err == nil {
		commit = strings.TrimSpace(string(outCommit))
	}

	cmdStatus := exec.Command("git", "status", "--porcelain")
	cmdStatus.Dir = fullPath
	outStatus, _ := cmdStatus.Output()
	status := "Clean"
	if len(strings.TrimSpace(string(outStatus))) > 0 {
		status = "Dirty"
	}

	return branch, commit, status
}

func generateDashboardHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
		return
	}

	repoPath := ".."

	submodules, err := getDashboardSubmodules(repoPath)
	if err != nil {
		http.Error(w, "Failed to get submodules", http.StatusInternalServerError)
		return
	}

	sort.Strings(submodules)

	var content strings.Builder
	content.WriteString("# Omni-Workspace Submodule Dashboard\n\n")
	content.WriteString(fmt.Sprintf("*Last updated: %s*\n\n", time.Now().Format("2006-01-02 15:04:05")))

	if len(submodules) == 0 {
		content.WriteString("No submodules detected in `.gitmodules`.\n")
	} else {
		content.WriteString("| Submodule | Branch | Commit | Status |\n")
		content.WriteString("| --- | --- | --- | --- |\n")

		for _, sm := range submodules {
			branch, commit, status := getRepoStatus(repoPath, sm)
			content.WriteString(fmt.Sprintf("| `%s` | `%s` | `%s` | %s |\n", sm, branch, commit, status))
		}
	}

	err = os.WriteFile(repoPath+"/SUBMODULE_DASHBOARD.md", []byte(content.String()), 0644)
	if err != nil {
		http.Error(w, "Failed to write dashboard file", http.StatusInternalServerError)
		return
	}

	resp := DashboardResponse{
		Success: true,
		Message: "Dashboard generated successfully",
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(resp)
}
